types/wlr_surface.c

```diff
@@ -365,12 +372,13 @@ static void surface_apply_damage(struct wlr_surface *surface) {
 		}
 	}
 
-	struct wlr_client_buffer *buffer =
-		wlr_client_buffer_import(surface->renderer, resource);
+	struct wlr_client_buffer *buffer = wlr_client_buffer_create(
+			surface->current.buffer, surface->renderer, resource);
 	if (buffer == NULL) {
 		wlr_log(WLR_ERROR, "Failed to upload buffer");
 		return;
 	}
+	wlr_buffer_unlock(surface->current.buffer);
```

Comment:
We should reset `surface->current.buffer` to NULL here, to avoid having someone read a free'd pointer.